### PR TITLE
Use TargetRef for Authorization Policy Enforcement

### PR DESCRIPTION
**Pull request by whitneygriffith**
**Please provide a description of this PR:**

Part of #46360 

This PR updates istio to fetch Authorization Policies to apply to a given waypoint based on targetRef field.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [x] Policies and Telemetry

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

**TODOs:**
- [x] Update tests 
- [x] Update release notes

### CODE REVIEW

**Comment by keithmattix** [comment]
Part of https://github.com/istio/istio/issues/46360

**Comment by whitneygriffith** [comment]
Blocked by #2888

**Comment by whitneygriffith** [comment]
This PR is blocked by [Define targetRef proto](https://github.com/istio/api/pull/2888)

File: pilot/pkg/model/authorization.go
Comment (keithmattix) [inline pilot/pkg/model/authorization.go:70]: I don't love the extra additions to the method signature here; can you put the selection criteria in some kind of `nodeInfo`data structure?

File: pilot/pkg/model/authorization.go
Comment (keithmattix) [inline pilot/pkg/model/authorization.go:97]: No need for extra assignment here right?

**Review by keithmattix — CHANGES_REQUESTED** [review]
Looks generally correct but needs test

File: pilot/pkg/model/authorization.go
Comment (keithmattix) [reply pilot/pkg/model/authorization.go]: Ahh I see; maybe a comment indicating this is for workloadSelectors would be helpful? Maybe even a TODO to remove this logic once we start disallowing workloadSelector for waypoints?

Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go]: Is it clearer now?

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [inline pilot/pkg/model/authorization.go:68]: @keithmattix  is this what you had in mind when you said to pass the proxy info as a struct?

Comment (keithmattix) [reply pilot/pkg/model/authorization.go:68]: Yeah something like that! Just a data structure to hold the matching criteria

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [inline pilot/pkg/model/authorization.go:92]: I am concerned that the current logic of looping through root ns and workload ns may end up applying a policy in the root ns for a waypoint that isn't in the root ns. 

Do I need to prevent this here? _By refactoring waypoint logic out of the ns loop or adding a check for if the root ns is not equal to the waypoint ns._ @keithmattix

Comment (keithmattix) [reply pilot/pkg/model/authorization.go:92]: Yeah waypoints don't care if a policy exists in the root namespace. I believe L85-87 shouldn't apply to waypoints

File: pilot/pkg/model/authorization.go
Comment (jaellio) [inline pilot/pkg/model/authorization.go:77]: Does `selectorAuthorizationPolicies` in pkg/serviceregistry/kube/controller/authorization.go also need to be updated to handle `targetRef`? If so, will this be addressed in a follow-up PR?

File: pilot/pkg/model/authorization.go
Comment (jaellio) [inline pilot/pkg/model/authorization.go:95]: Can we assume `targetRef.Group` is valid at this point?

File: pilot/pkg/model/authorization.go
Comment (jaellio) [inline pilot/pkg/model/authorization.go:91]: We could consider adding a log line here to indicate the target is a waypoint proxy, but does not have a `targetRef`.

File: pilot/pkg/model/push_context.go
Comment (jaellio) [inline pilot/pkg/model/push_context.go:1221]: Could we add this info as a function comment?

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go
Comment (jaellio) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:69]: It looks like this has been addressed by this PR. Can we remove this TODO?

File: pkg/fuzz/README.md
Comment (jaellio) [inline pkg/fuzz/README.md:28]: Address TODO

File: pilot/pkg/model/context.go
Comment (jaellio) [inline pilot/pkg/model/context.go:766]: Remove TODO

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:95]: Yes, we are assuming that the policy is a valid policy at this point. This is ensuring the waypoint that made this requests matches the referent in the targetRef

File: pkg/fuzz/README.md
Comment (whitneygriffith) [inline pkg/fuzz/README.md:24]: Should proxyInfo be fuzzed?

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:92]: Good catch! I added the log line.

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:77]: Great find! I see that there is workload selector logic here. I need to understand how `selectorAuthorizationPolicies` is used before I can say we want to update it to handle targetRef for sure.

Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:77]: From what I gather I believe we should update this since this is a function used to precompute **only** ambient workload information. 

What do you think? @keithmattix

Comment (keithmattix) [reply pilot/pkg/model/authorization.go:77]: Yeah this is a great catch! Indeed, that function needs to be updated to handle targetRef as well

**Comment by whitneygriffith** [comment]
Resolves #46847

File: pilot/pkg/model/authorization.go
Comment (keithmattix) [reply pilot/pkg/model/authorization.go:75]: Actually, these only deal with the ztunnel authz policies I think. So it may not need a change

Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:75]: I wouldn't make a change to it in this PR. If needed in the future we can do a follow up PR.

File: pilot/pkg/model/authorization.go
Comment (keithmattix) [reply pilot/pkg/model/authorization.go:68]: Yeah I think you can probably use `workloadSelectionOpts` instaed

Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:68]: I will keep it as we may need additional workload selection options in the future.

**Review by keithmattix — COMMENTED** [review]
RequestAuthentication tests are failing because they rely on label selector for authorization policy; you'll need to update the YAML templates [here](https://github.com/istio/istio/blob/master/tests/integration/security/policy_attachment_only/testdata/requestauthn/gateway-jwt.yaml.tmpl#L35) and [here](https://github.com/istio/istio/blob/master/tests/integration/security/testdata/requestauthn/gateway-jwt.yaml.tmpl#L35)

**Review by keithmattix — APPROVED** [review]
Content of the PR looks good. For a core resource like AuthorizationPolicy, I feel like it would be beneficial to have at least 1 integration test. I'll defer to security and test/release maintainers there though

**Comment by whitneygriffith** [comment]
> Content of the PR looks good. For a core resource like AuthorizationPolicy, I feel like it would be beneficial to have at least 1 integration test. I'll defer to security and test/release maintainers there though

I can follow up with an integration test after this merges.

File: releasenotes/notes/46847.yaml
Comment (ericvn) [inline releasenotes/notes/46847.yaml:8]: Not sure if we should backtick AuthorizationPolicy.

File: pilot/pkg/networking/plugin/authz/authorization.go
Comment (ericvn) [inline pilot/pkg/networking/plugin/authz/authorization.go:57]: Nit: You added some blank lines within the code, but haven't done it anywhere else within this PR.

File: pilot/pkg/model/authorization.go
Comment (howardjohn) [inline pilot/pkg/model/authorization.go:93]:
```suggestion
			switch getPolicyMatcher(gvk.AuthorizationPolicy, config.Name, selectionOpts, spec) {
			case policyMatchSelector:
				selector := spec.GetSelector().GetMatchLabels()
```

File: pilot/pkg/model/authorization.go
Comment (howardjohn) [inline pilot/pkg/model/authorization.go:99]:
```suggestion
				continue
```
this does nothing (?)

Same with L101

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/authorization.go
Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:93]: If I don't declare `var selector labels.Instance` separately the selector will be inferred as `map[string]string` type from the `GetMatchLabels()` result and  `SubsetOf()` wouldn't be available for L94

Comment (howardjohn) [reply pilot/pkg/model/authorization.go:93]: you can do `selector := labels.Instance(spec.GetSelector().GetMatchLabels())` or `var selector labels.Instance = spec.GetSelector().GetMatchLabels()`

Comment (whitneygriffith) [reply pilot/pkg/model/authorization.go:93]: Thanks!

**Pull request merged**
